modules/xfem/src/base/XFEM.C

```diff
@@ -1507,3 +1557,162 @@ XFEM::getXFEMqRuleOnSurface(std::vector<Point> & intersection_points, std::vecto
+                             const NumericVector<Number> & current_solution,
+                             const NumericVector<Number> & old_solution)
+{
+  const std::vector<MooseVariable *> & vars = sys.getVariables(0);
+  _stored_solution_scratch.clear();
+  _stored_solution_scratch.reserve(2 * vars.size());
+  _stored_solution_dofs_scratch.clear();
+  _stored_solution_dofs_scratch.reserve(vars.size());
+  SubdomainID sid = elem_to_store_from->subdomain_id();
+  for (auto var : vars)
+  {
+    if (!var->isNodal())
+    {
+      const std::set<SubdomainID> & var_subdomains = sys.getSubdomainsForVar(var->number());
+      if (var_subdomains.find(sid) != var_subdomains.end())
+      {
+        unsigned int elem_dof = elem_to_store_from->dof_number(sys.number(), var->number(), 0);
+        _stored_solution_dofs_scratch.push_back(elem_dof);
+      }
+    }
+  }
+
+  for (auto dof : _stored_solution_dofs_scratch)
+    _stored_solution_scratch.push_back(current_solution(dof));
+
+  for (auto dof : _stored_solution_dofs_scratch)
+    _stored_solution_scratch.push_back(old_solution(dof));
+
+  stored_solution[elem_to_store_to->unique_id()] = _stored_solution_scratch;
+  stored_solution[elem_to_store_to->unique_id()].shrink_to_fit();
+}
+
+void
+XFEM::setNodeSolution(SystemBase & sys,
+                      const std::map<unique_id_type, std::vector<Real>> & stored_solution,
+                      NumericVector<Number> & current_solution,
+                      NumericVector<Number> & old_solution)
+{
+  const std::vector<MooseVariable *> & vars = sys.getVariables(0);
+  _stored_solution_dofs_scratch.reserve(vars.size());
+
+  for (auto nit = stored_solution.begin();
+       nit != stored_solution.end(); ++nit)
+  {
+    _stored_solution_dofs_scratch.clear();
+    Node* node = getNodeFromUniqueID(nit->first);
+    const std::vector<Real> & node_solution = nit->second;
+    if (node->processor_id() == _mesh->processor_id())
+    {
+      const std::set<SubdomainID> & sids = _moose_mesh->getNodeBlockIds(*node);
+
+      for (auto var : vars)
+      {
+        if (var->isNodal())
+        {
+          const std::set<SubdomainID> & var_subdomains = sys.getSubdomainsForVar(var->number());
+          std::set<SubdomainID> intersect;
+          set_intersection(var_subdomains.begin(), var_subdomains.end(), sids.begin(), sids.end(),
+                           std::inserter(intersect, intersect.begin()));
+          if (var_subdomains.empty() || !intersect.empty())
+          {
+            unsigned int node_dof = node->dof_number(sys.number(), var->number(), 0);
+            _stored_solution_dofs_scratch.push_back(node_dof);
+          }
+        }
+      }
+    }
+    for (unsigned int i = 0; i < _stored_solution_dofs_scratch.size(); ++i)
+    {
+      current_solution.set(_stored_solution_dofs_scratch[i], node_solution[i]);
+      old_solution.set(_stored_solution_dofs_scratch[i], node_solution[_stored_solution_dofs_scratch.size() + i]);
```

Comment:
Yeah, I see it now. There's a little voodoo on this line with your storage indexing. It probably should be documented.